src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Validation.cs

```diff
@@ -0,0 +1,728 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+//
+// This file contains workhorse methods for performing validation of UTF-8 byte sequences.
+//
+
+namespace System.Buffers.Text
+{
+    internal static partial class Utf8Util
+    {
+        // This method will consume as many ASCII bytes as it can using fast vectorized processing, returning the number of
+        // consumed (ASCII) bytes. It's possible that the method exits early, perhaps because there is some non-ASCII byte
+        // later in the sequence or because we're running out of input to search. The intent is that the caller *skips over*
+        // the number of bytes returned by this method, then it continues data processing from the next byte.
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static unsafe int ConsumeAsciiBytesVectorized(ref byte buffer, int length)
+        {
+            // Only allow vectorization if vector operations are hardware-accelerated
+            // and vectors are at least 128 bits in length.
+            if (Vector.IsHardwareAccelerated && (Vector<byte>.Count >= 2 * sizeof(ulong)))
+            {
+                // Need to pin buffer so the GC doesn't cause it to be misaligned when we're
+                // reading vectors from it.
+                fixed (byte* pbBuffer = &buffer)
+                {
+                    int offsetAtWhichVectorIsAligned;
+                    {
+                        int referenceMisalignment = (int)pbBuffer % Vector<byte>.Count;
+                        offsetAtWhichVectorIsAligned = (referenceMisalignment == 0) ? 0 : (Vector<byte>.Count - referenceMisalignment);
+                    }
+
+                    if (length - offsetAtWhichVectorIsAligned < (Vector<byte>.Count * 2))
+                    {
+                        // After alignment there's not enough data left over to warrant setting
+                        // up the vectorization code path.
+                        return 0;
+                    }
+
+                    int numBytesConsumed = 0;
+
+                    // Consume QWORDs or DWORDs until we're aligned with where we can start vectorization.
+                    // It's ok if we duplicate a little bit of work by having a single QWORD or DWORD read
+                    // overlap with the vector read area. It's not worth the overhead of checking for this
+                    // condition and avoiding the duplicate work.
+
+                    if (IntPtr.Size >= sizeof(ulong))
+                    {
+                        while (numBytesConsumed < offsetAtWhichVectorIsAligned)
+                        {
+                            if (!QWordAllBytesAreAscii(Unsafe.ReadUnaligned<ulong>(&pbBuffer[numBytesConsumed])))
```

Comment:
You know you can just reinterpret cast Span<byte> to Vector<byte> without having to use unsafe code/pointers. See https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/SpanHelpers.byte.cs#L474 for example.

Reply:
Yes. However, this is an unaligned read, and the JITter produces optimal assembly for `Unsafe.ReadUnaligned<T>(void*)` but _not_ for `Unsafe.ReadUnaligned<T>(ref Unsafe.Add(...))`. I'm taking advantage of the fact that the buffer is already pinned in order to let me call the optimal overload.